scrapscript.py

```diff
@@ -521,68 +544,94 @@ def parse_unary(tokens: Peekable, p: float) -> "Object":
             l = Bytes(base64.b16decode(token.value))
         else:
             raise ParseError(f"unexpected base {base!r} in {token!r}")
+        object.__setattr__(l, "source_extent", token.source_extent)
         return l
     elif isinstance(token, StringLit):
-        return String(token.value)
+        return make_source_annotated_object(String, token.source_extent, token.value)
     elif token == Operator("..."):
         try:
             if isinstance(tokens.peek(), Name):
-                return Spread(next(tokens).value)
+                spread_variable = next(tokens)
+                return make_source_annotated_object(
+                    Spread, token.source_extent.coalesce(spread_variable.source_extent), spread_variable.value
+                )
             else:
-                return Spread()
+                return make_source_annotated_object(Spread, token.source_extent)
         except StopIteration:
             return Spread()
     elif token == Operator("|"):
+        pipe_source_extent = token.source_extent
         expr = parse_binary(tokens, PS["|"].pr)  # TODO: make this work for larger arities
         if not isinstance(expr, Function):
             raise ParseError(f"expected function in match expression {expr!r}")
-        cases = [MatchCase(expr.arg, expr.body)]
+        match_case = make_source_annotated_object(
+            MatchCase, pipe_source_extent.coalesce(expr.source_extent), expr.arg, expr.body
+        )
+        cases = [match_case]
         while True:
             try:
                 if tokens.peek() != Operator("|"):
                     break
             except StopIteration:
                 break
-            next(tokens)
+            pipe_source_extent = next(tokens).source_extent
             expr = parse_binary(tokens, PS["|"].pr)  # TODO: make this work for larger arities
             if not isinstance(expr, Function):
                 raise ParseError(f"expected function in match expression {expr!r}")
-            cases.append(MatchCase(expr.arg, expr.body))
-        return MatchFunction(cases)
+            match_case = make_source_annotated_object(
+                MatchCase, pipe_source_extent.coalesce(expr.source_extent), expr.arg, expr.body
+            )
+            cases.append(match_case)
+        cases_source_extents = [case_branch.source_extent for case_branch in cases]
+        return make_source_annotated_object(
+            MatchFunction,
+            reduce(
```

Comment:
Nit: we already have a loop and can join them in the loop instead of using `reduce`